Rubberduck.Parsing/VBA/COMReferenceManagerBase.cs

```diff
@@ -0,0 +1,269 @@
+                return _lastRunLoadedReferences;
+            }
+        }
+
+        private bool _lastRunUnloadedReferences;
+        public bool LastRunUnloadedReferences
+        {
+            get
+            {
+                return _lastRunUnloadedReferences;
+            }
+        }
+
+        private readonly HashSet<ReferencePriorityMap> _projectReferences = new HashSet<ReferencePriorityMap>();
+        public ICollection<ReferencePriorityMap> ProjectReferences
+        {
+            get
+            {
+                return _projectReferences.ToHashSet();
+            }
+        }
+
+
+        protected abstract void LoadReferences(List<IReference> referencesToLoad, ConcurrentBag<IReference> unmapped, CancellationToken token);
+
+
+        public void SyncComReferences(IReadOnlyList<IVBProject> projects, CancellationToken token)
+        {
+            _lastRunLoadedReferences = false;
+            _lastRunUnloadedReferences = false;
+
+            var unmapped = new ConcurrentBag<IReference>();
+
+            var referencesToLoad = GetReferencesToLoadAndSaveReferencePriority(projects);
+
+            _state.OnStatusMessageUpdate(ParserState.LoadingReference.ToString()); //Why do we use this method to change the displayed state?
+
+            if (referencesToLoad.Any())
+            {
+                _lastRunLoadedReferences = true;
+                LoadReferences(referencesToLoad, unmapped, token);
+            }
+
+            var notMappedReferences = NonMappedReferences(projects);
+            foreach (var item in notMappedReferences)
+            {
+                unmapped.Add(item);
+            }
+
+            if (unmapped.Any())
+            {
+                _lastRunUnloadedReferences = true;
+                foreach (var reference in unmapped)
+                {
+                    UnloadComReference(reference, projects);
+                }
+            }
+        }
+
+        private List<IReference> GetReferencesToLoadAndSaveReferencePriority(IReadOnlyList<IVBProject> projects)
+        {
+            var referencesToLoad = new List<IReference>();
+
+            foreach (var vbProject in projects)
+            {
+                var projectId = QualifiedModuleName.GetProjectId(vbProject);
+                var references = vbProject.References;
+
+                // use a 'for' loop to store the order of references as a 'priority'.
+                // reference resolver needs this to know which declaration to prioritize when a global identifier exists in multiple libraries.
+                for (var priority = 1; priority <= references.Count; priority++)
+                {
+                    var reference = references[priority];
+                    if (reference.IsBroken)
+                    {
+                        continue;
+                    }
+
+                    // skip loading Rubberduck.tlb (GUID is defined in AssemblyInfo.cs)
+                    if (reference.Guid == rubberduckGUID)
```

Comment:
@Hosch250 this was merely moved from where it was and left as-is. I like that there's this `todo` comment to remind us that COM-collecting Rubberduck.tlb has been (and still can be) a problem. Perhaps issue #2789 should be referenced here though.